### DISCUSSION TITLE
Virtual Race Course and Gates in OSD

### DISCUSSION THREAD

**Discussion by robotgoat** [iNavFlight/inav#10066] [Ideas]
I've gotten addicted to converting regular EDF planes to Inav and flying them FPV in a sporty manner just like VLOS pilots do. While enjoying this experience, I thought how cool would it be to create a race course in Inav in the same vein you can currently create mission waypoints. 

The gist of this race course feature would involve the creation of a new tab or subtab within the "Mission Control" tab. You would place waypoints, which would act as race gates in this case, that are defined by a GPS altitude, longitude and latitude, and diameter from that point. This gate would then appear in the OSD as an actual circle on the OSD that the pilot would have to fly through to make that checkpoint. The timer for the circuit, or segment, would begin when Inav detects that the airplane has passed through the starting gate and stop when the pilot resets the course or passes the finish gate. 

Is this feature feasible given the current hardware limitations of F405 boards and if so, how easy or hard would it be to start implementing? I would like to start implementing, but need to understand the Inav codebase better. Please disregard if this is too silly of an idea. Thanks!

**Comment by Scavanger** [comment]
Two thoughts on this, although the idea does have something to it:

- It's not enough to simply place the gate somewhere, you also have to calculate what the gate looks like when you fly towards it from the side/above etc., which could be tight on the microprecessors

- The INAV OSD is letter-based, i.e. you have a resolution of 30 x 13 (analogue NTSC) to 60x22 (DJI HD), and the gate must not overlay any OSD elements. So it may well be that most of the gate is simply not visible and the resolution is so coarse that it is difficult to fly through it precisely.

**Reply by robotgoat** [reply]
Thanks for the insight! I had feared the current STM chips would be a possible limitation. Perhaps this will be a more realistic idea later on a different set of hardware. As for displaying the gate, perhaps it doesn't have to be an actual gate, rather it can be indicated by a chevron or pointer with an altitude ticker next to it? It would be the pilots' responsibility to have a metal picture of the race course like identifying physical markers on the ground near each gate/checkpoint.

**Comment by andreelb** [comment]
To add to this discussion instead of making a new thread. I envision it being just like waypoints in games. You have the active gate/WP and you have the next gate/WP. Passing through a gate removes it, makes the next gate active and starts showing the gate after that.

Doesn't need to be a full graphical gate - just a point on the screen. if the point is off screen, or outside a certain area available to display waypoints, just show a chevron/arrow showing in what direction it is in. then you'd turn until you see it on screen.

See my post in INAV group on FB:

https://www.facebook.com/groups/INAVFixedWing/posts/2389469487921502/

**Comment by sensei-hacker** [comment]
Note that as Andreelb hinted at, this is *already* available mostly, as best as it can be with current OSDs.

You can set up waypoints as if for a waypoint mission and have those displayed on the OSD.

**Comment by jasc76** [comment]
I really like the idea !!
Also this „gamification“ approach !!
A simple point on the osd could be sufficient I guess. Also no direction check or grafics. It should show if you are too high or too low (for „air gates“ or lets call it points) for the current point. So every point should have a low, high and a max distance value.
Then a number of laps and a start/finish point

of course setting this up in a free field is possible, most useful could be, to set them together with landmarks. But of course not needed. Also the min height is optional.

I think its best suited on current HD OSD (Hdz and WS) (F DJI) and well maybe this is a feature thats just nearly usable on anal-log but well, times are changing, stuff is getting outdated… (f3/f1 users out there running Inav 2… more power to you 👍🏻)

oh oh, and and ex/import function to have everyone quickly on the same course
And race stats (current lap, best lap, total time, laps to go)
